### PR TITLE
KJS IR: Add string concatenation lowering and fixes KT-39891

### PR DESCRIPTION
**Pull request by awenger**
The precision loss in [KT-39891](https://youtrack.jetbrains.com/issue/KT-39891) is caused by `Long::valueOf`, which is casting the Long value to a JavaScript Number before it concatenates it with the string. This is happening with ordinary `String::plus`, as well as string concatenations (`"abc $longValue"`). This PR fixes it with two new lowering steps:
- Lowering step for string concatenation -> chain of `String::plus`
- Lowering step that fixes lost precision caused by `String::plus(Long)`. This step converts such Long values to String before passing them to `String::plus`

### CODE REVIEW

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/JsLoweringPhases.kt [hunk 1]
```diff
@@ -660,6 +662,20 @@ private val blockDecomposerLoweringPhase = makeBodyLoweringPhase(
     prerequisite = setOf(typeOperatorLoweringPhase, suspendFunctionsLoweringPhase)
 )
 
+private val stringConcatenationLowering = makeBodyLoweringPhase(
+    ::StringConcatenationLowering,
+    name = "StringConcatenationLowering",
+    description = "Lowers String concatenation to a chain of String::plus",
+    prerequisite = setOf(blockDecomposerLoweringPhase)
```
Comment (awenger) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/JsLoweringPhases.kt:669]: The `blockDecomposerLoweringPhase` step creates String concatenations, so it needs to run first

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/JsLoweringPhases.kt [hunk 2]
```diff
@@ -675,7 +691,8 @@ private val primitiveCompanionLoweringPhase = makeBodyLoweringPhase(
 private val callsLoweringPhase = makeBodyLoweringPhase(
     ::CallsLowering,
     name = "CallsLowering",
-    description = "Handle intrinsics"
+    description = "Handle intrinsics",
+    prerequisite = setOf(stringPlusLongLowering)
```
Comment (awenger) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/JsLoweringPhases.kt:695]: This one lowers the `String::plus` to `jsPlus`, so the fix for Longs needs to happen first

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/StringConcatenationLowering.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        if (!expression.type.isNullable() && typeToString != null) {
+            return JsIrBuilder.buildCall(typeToString).apply {
+                dispatchReceiver = expression
+            }
+        }
+
+        return JsIrBuilder.buildCall(context.ir.symbols.extensionToString).apply {
+            extensionReceiver = expression
+        }
+    }
```
Comment (awenger) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/StringConcatenationLowering.kt:66]: Is there maybe already some helper/utility method somewhere in the codebase to add a `toString` call properly?

**Comment by broadwaylamb** [comment]
Thank you @awenger!

We've decided to do some refactoring and then implement the Long conversion on top of that.

The issue has been fixed in 42c213d, so I'll close the PR.

**Comment by awenger** [comment]
Thanks allot @broadwaylamb and great that it is fixed now

**Pull request closed without merging**